### ISSUE TITLE
KDE Konsole doesn't seem to respond to "get screen size" escape sequence

### ISSUE THREAD

**Issue by arcivanov** [pfalcon/picotui#48]
> X11 Konsole/bash on Fedora 33 with Python 3.9.1

Firstly, thank you for the project, it reminds me of the good-old Borland Turbo Vision, which was absolutely epic.

However, I just started writing a prototype and the most basic features seem not to work. I guess it's something specific with my setup (Python 3.9?), but I'm still going to report it in case the problem is unknown and global.

Just for the reference, other console tools like `bpytop` and `s-tui` work perfectly.

1. Mouse is not working at all.
2. TTY size is not retrieved.

I suspect it's all related to console comms but it's hard to say what's going on.

As far as TTY size I always end up going this route with `res` being `[]`:
https://github.com/pfalcon/picotui/blob/7a26e620329749e60b2b9a3969012cee1fc29294/picotui/screen.py#L144-L146
`bpytop` uses `os.get_terminal_size()` which seems to work perfectly:
https://github.com/aristocratos/bpytop/blob/2fbdc35125ad5d9a9f613736f194a33ec69bb14b/bpytop.py#L651

Anybody else has issues with basic functionality like that? I'm running code virtually identical to the https://github.com/pfalcon/picotui/blob/master/example_menu.py

**Comment by pfalcon** [comment]
> I suspect it's all related to console comms but it's hard to say what's going on.

Exactly, as you don't say what you use, it's hard to say what's going on. Please see https://www.chiark.greenend.org.uk/~sgtatham/bugs.html .

**Comment by arcivanov** [comment]
First of all, it's not even a bug report (yet). I'm trying to gauge whether there are other people that might have the same massive issue of "nothing working right". Maybe there is a common quirk that is known to all, as in maybe everybody knows that the project doesn't work on 3.9 or that the examples in project root are broken and only the examples in the `examples` directory do, or something else along those lines. There is no documentation to speak of, hence the broad question.

Secondly, "you don't say what you use" is not true. I said I'm on Fedora 33, Python 3.9.1, KDE Konsole over X11 and bash (over actual tty). I am using the latest version of picotui.

Thirdly, I even posted specific lines in terminal size query code that fail, returning default terminal size path. That, generally, shouldn't happen (ever?), so there must be something fundamentally wrong.

**Comment by arcivanov** [comment]
So this monkey-patching workaround for the terminal sizing fixes things somewhat:

```python
from picotui import screen


class Screen(screen.Screen):
    @classmethod
    def screen_size(cls):
        return os.get_terminal_size(sys.stdout.fileno())


screen.Screen = Screen

from picotui.context import Context
from picotui.menu import (C_WHITE,
                          C_BLUE,
                          C_B_WHITE,
                          C_BLACK,
                          ACTION_CANCEL,
                          KEY_F9,
                          ACTION_OK,
                          WMenuBox, WMenuBar, Screen)
from picotui.widgets import Dialog, WButton, WLabel, WListBox, WDropDown
...
```

**Comment by pfalcon** [comment]
>  said I'm on Fedora 33, Python 3.9.1, KDE Konsole over X11

Right, please consider a brain-fart. But for modern short attention-span humans, like myself, I'm going to edit the actual description to contain that info.

**Comment by pfalcon** [comment]
As a quick reply, my work environment is Ubuntu MATE with MATE Terminal (1.24 as of now, but different version over last few years). Also tested with Gnome Terminal (of which MATE Terminal is a fork). I'll try KDE Konsole as time permits. In the meantime, please post what you get in it with:

```
$ env|grep TERM
COLORTERM=truecolor
TERM=xterm-256color
```

**Comment by pfalcon** [comment]
> maybe everybody knows that the project doesn't work on 3.9

I wouldn't err on the side of the of Python version, I'd err on what terminal setup you and/or in which terminal emulator. Just to confirm, I quick ran example_widgets.py with:
~~~
$ python3.9 --version
Python 3.9.1
~~~
and don't see any issues.

> or that the examples in project root are broken and only the examples in the examples directory do, or something else along those lines.

Chances are that situation would be opposite - examples in the root is tried all the time, while stuff at `examples/` might bitrot.

> Thirdly, I even posted specific lines in terminal size query code that fail, returning default terminal size path. That, generally, shouldn't happen (ever?)

Terminal size querying relies on the particular Xterm escape sequence. MATE Terminal handles that sequence in a way compatible with picotui (speaking conservatively).

**Comment by pfalcon** [comment]
> So this monkey-patching workaround for the terminal sizing fixes things somewhat:

That doesn't tell me much, I would need to see a diff.

**Comment by pfalcon** [comment]
Confirming Xterm mouse not working with Konsole 19.12.3.

**Comment by pfalcon** [comment]
Screen size detection also doesn't seem to work, yeah.

**Comment by pfalcon** [comment]
My guess is that Konsole wants some escape sequence to enable all the "extra" features like mouse input and size reporting...

**Comment by arcivanov** [comment]
Please see #49 for terminal size fix/workaround. The call is available in all versions of Python that aren't EOLed. I maintained the alternative for all platforms where it's not available.

**Comment by pfalcon** [comment]
Ok, mouse is working in Konsole in the latest master.

**Comment by pfalcon** [comment]
Suggested further plan re: screen size issue:

1. Review https://www.x.org/docs/xterm/ctlseqs.pdf and compare with what picotui sends.
2. If everything's ok on picotui side, figure out why Konsole doesn't respond to that escape. There're multiple ways to do that, the ultimate is to look at the source.
3. Imagining that Konsole simply doesn't implement it, submit an upstream issue on it ("Konsole doesn't conform to Xterm terminal spec it advertises as implementing").
4. Only after that can consider workarounds on picotui side.

**Comment by ethael** [comment]
The same "resize" problem exists when using "st" simple terminal from suckless:  https://st.suckless.org
Proposed patch does resolve the problem on "st" too.
